File: 2558_take_gifts_from_the_richest_pile.java
class Solution {
    public long pickGifts(int[] gifts, int k) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        long totalSum = 0;
        for (int gift : gifts) {
            maxHeap.add(gift);
            totalSum += gift;
        }
        while (k-- > 0) {
            int largest = maxHeap.poll();
            totalSum -= (largest - Math.sqrt(largest));
            maxHeap.add((int) Math.sqrt(largest));
        }
        return totalSum;
    }
}

File: 2558_take_gifts_from_the_richest_pile.cpp
class Solution {
public:
    long long pickGifts(vector<int>& gifts, int rounds) {
        priority_queue<int> maxHeap;
        long long totalSum = 0;
        for (int gift : gifts) {
            maxHeap.push(gift);
            totalSum += gift;
        }
        while (rounds--) {
            int largest = maxHeap.top();
            totalSum -= (largest - sqrt(largest));
            maxHeap.pop();
            maxHeap.push(sqrt(largest));
        }
        return totalSum;
    }
};
